File: typescript/packages/core/src/core/gsheets/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GSheetsAccessor } from '../../accessor/gsheets.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { globToModifiedRange } from '../google/date_glob.ts'
import { listAllFiles } from '../google/drive.ts'
import { makeFilename } from '../../resource/gsheets/sheet_entry.ts'
import { stripSlash } from '../../util/slash.ts'

const MIME = 'application/vnd.google-apps.spreadsheet'

export async function readdir(
  accessor: GSheetsAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  const prefix = path.prefix
  const modifiedRange = path.pattern ? globToModifiedRange(path.pattern) : null
  const raw = path.pattern ? path.directory : path.original
  let p = raw
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  const virtualKey = key !== '' ? `${prefix}/${key}` : prefix !== '' ? prefix : '/'

  if (key === '') return [`${prefix}/owned`, `${prefix}/shared`]

  if (key !== 'owned' && key !== 'shared') {
    const e = new Error(`ENOENT: ${path.original}`) as Error & { code: string }
    e.code = 'ENOENT'
    throw e
  }

  if (index !== undefined && modifiedRange === null) {
    const cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) return cached.entries
  }

  const files = await listAllFiles(accessor.tokenManager, {
    mimeType: MIME,
    modifiedAfter: modifiedRange ? modifiedRange[0] : null,
    modifiedBefore: modifiedRange ? modifiedRange[1] : null,
  })
  const isOwned = key === 'owned'
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const f of files) {
    const owners = f.owners ?? []
    const firstOwner = owners[0] ?? {}
    const fileOwned = firstOwner.me === true
    if (fileOwned !== isOwned) continue
    const filename = makeFilename(f.name, f.id, f.modifiedTime ?? '')
    const sizeRaw = f.size ?? f.quotaBytesUsed ?? '0'
    const sizeNum = Number.parseInt(sizeRaw, 10)
    entries.push([
      filename,
      new IndexEntry({
        id: f.id,
        name: f.name,
        resourceType: 'gsheets/file',
        remoteTime: f.modifiedTime ?? '',
        vfsName: filename,
        size: Number.isFinite(sizeNum) && sizeNum > 0 ? sizeNum : null,
      }),
    ])
    names.push(`${prefix}/${key}/${filename}`)
  }

  if (index !== undefined) {
    if (modifiedRange !== null) {
      for (const [name, entry] of entries) {
        await index.put(`${virtualKey}/${name}`, entry)
      }
    } else {
      await index.setDir(virtualKey, entries)
    }
  }
  return names
}

File: python/mirage/core/gslides/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.gslides import GSlidesAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.google.date_glob import glob_to_modified_range
from mirage.core.google.drive import GoogleFileSuffix, list_all_files
from mirage.resource.gslides.slide_entry import make_filename
from mirage.types import PathSpec
from mirage.utils.errors import enoent

MIME = "application/vnd.google-apps.presentation"


def is_dir_name(child: str) -> bool:
    # readdir emits only folders and rendered *.gslide.json files.
    return not child.endswith(GoogleFileSuffix.GSLIDE.value)


async def readdir(
    accessor: GSlidesAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    modified_range = None
    if isinstance(path, PathSpec):
        prefix = path.prefix
        if path.pattern:
            modified_range = glob_to_modified_range(path.pattern)
        path = path.directory if path.pattern else path.original
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    virtual_key = prefix + "/" + key if key else prefix or "/"

    if not key:
        return [f"{prefix}/owned", f"{prefix}/shared"]

    if key not in ("owned", "shared"):
        raise enoent(virtual)

    if index is not None and not modified_range:
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries

    range_kwargs: dict[str, str] = {}
    if modified_range:
        range_kwargs["modified_after"] = modified_range[0]
        range_kwargs["modified_before"] = modified_range[1]
    files = await list_all_files(accessor.token_manager,
                                 mime_type=MIME,
                                 **range_kwargs)
    is_owned = key == "owned"
    entries = []
    for f in files:
        owners = f.get("owners", [])
        first_owner = owners[0] if owners else {}
        file_owned = first_owner.get("me", False)
        if file_owned != is_owned:
            continue
        filename = make_filename(f["name"], f["id"], f.get("modifiedTime", ""))
        entry = IndexEntry(
            id=f["id"],
            name=f["name"],
            resource_type="gslides/file",
            remote_time=f.get("modifiedTime", ""),
            vfs_name=filename,
            size=int(f.get("size") or f.get("quotaBytesUsed") or 0) or None,
        )
        entries.append((filename, entry))

    if index is not None:
        if modified_range:
            for name, entry in entries:
                await index.put(f"{virtual_key}/{name}", entry)
        else:
            await index.set_dir(virtual_key, entries)
    return [f"{prefix}/{key}/{name}" for name, _ in entries]
